turbine/src/xdp.rs

```diff
@@ -0,0 +1,183 @@
+// re-export since this is needed at validator startup
+pub use agave_xdp::set_cpu_affinity;
+#[cfg(target_os = "linux")]
+use {
+    agave_xdp::{
+        device::{NetworkDevice, QueueId},
+        load_xdp_program,
+        tx_loop::tx_loop,
+    },
+    crossbeam_channel::TryRecvError,
+    std::{sync::Arc, thread::Builder, time::Duration},
+};
+use {
+    crossbeam_channel::{Sender, TrySendError},
+    solana_ledger::shred,
+    std::{error::Error, net::SocketAddr, thread},
+};
+
+#[derive(Clone, Debug)]
+pub struct XdpConfig {
+    pub interface: Option<String>,
+    pub cpus: Vec<usize>,
+    pub zero_copy: bool,
+    pub channel_cap: usize,
+}
+
+impl XdpConfig {
+    // A nice round number
+    const DEFAULT_CHANNEL_CAP: usize = 1_000_000;
+}
+
+impl Default for XdpConfig {
+    fn default() -> Self {
+        Self {
+            interface: None,
+            cpus: vec![],
+            zero_copy: false,
+            channel_cap: Self::DEFAULT_CHANNEL_CAP,
+        }
+    }
+}
+
+impl XdpConfig {
+    pub fn new(interface: Option<impl Into<String>>, cpus: Vec<usize>, zero_copy: bool) -> Self {
+        Self {
+            interface: interface.map(|s| s.into()),
+            cpus,
+            zero_copy,
+            channel_cap: XdpConfig::DEFAULT_CHANNEL_CAP,
+        }
+    }
+}
+
+pub(crate) struct XdpSender {
+    senders: Vec<Sender<(Vec<SocketAddr>, shred::Payload)>>,
+}
+
+impl XdpSender {
+    #[inline]
+    pub(crate) fn try_send(
+        &self,
+        shred_index: u32,
```

Comment:
bikeshedding: I wonder if we could call this something besides `shred_index`. We happen to be using `shred_index` to fan out to the senders, but it could just as easily use the enumeration index that the legacy code uses

Reply:
renamed to sender_index